Add HRX (amdxdna) host-runtime backend and XRT-free xclbin packaging (#3347)
Xilinx/mlir-aie: python/utils/hostruntime/hrxruntime/README.md

> <!---//===- README.md --------------------------*- Markdown -*-===//

Reviewer: I'm very happy to see the documentation! We currently rely on two types of documentation:
* docstrings, which (for the most part) are rendered on the website https://xilinx.github.io/mlir-aie/dev/api/iron/
* markdown based docs served by the website: e.g., https://xilinx.github.io/mlir-aie/dev/programming_guide/iron_configuration/

In general, we have few embedded READMEs because they can be a bit hard to find for folks who install from wheels, etc. but do not clone source code. I noticed your docstrings don't include all the arg fields that the other docstrings do, and this README isn't available on the website (I don't think).

This could be a follow up, but I do think integrating HRX runtime documentation into the existing systems would be a very nice follow up.

Author: Have addressed in this PR. Can you please take a look?

Reviewer: Looks great, thank you!

//
// Copyright (C) 2026 Advanced Micro Devices, Inc.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//-->

> # Running MLIR-AIE IRON (Python) & C++ tests on the **HRX** runtime

Reviewer 2: ```suggestion
# Running IRON (Python) & C++ tests on the **HRX** runtime
```

This package (`aie.utils.hostruntime.hrxruntime`) is the **HRX** host-runtime

> backend for MLIR-AIE. It dispatches IRON designs and the C++ testbench on the

Reviewer 2: ```suggestion
backend for IRON. It dispatches IRON designs and the C++ testbench on the
```

| **HRX built** (`libhrx.so`) | See §2. This is the runtime we dispatch through; it exports `hrx_amdxdna_executable_create`, which builds the amdxdna XADX package internally. |
| **MLIR-AIE installed from *this* branch** | So `import aie` works, `aiecc` can build examples, and the wheel includes this `hrxruntime` package + the `IRON_RUNTIME` switch. See §3. |
| **A C++ toolchain + CMake ≥ 3.30** | Only for the C++ tests. (`pip install "cmake>=3.30"` into the venv if the system cmake is older.) |
| **Peano / llvm-aie** | Needed by `aiecc` to build the example `.xclbin`/`insts.bin` (same as the XRT flow). |
| **`xclbinutil`** (xclbin packaging) | `aiecc` packages the `.xclbin` with the tool named by **`AIE_XCLBINUTIL`** / **`--xclbinutil-path`** (explicit, deterministic — see §3b), falling back to a `PATH` lookup only if neither is set. Provided by XRT, **or** build the bundled XRT-free `hrx-xclbinutil` (§3b) — recommended on an XRT-free box. |
| XRT userspace (pyxrt) | **Optional.** Only needed for the XRT baseline. HRX itself does not need it. |

The NPU + driver + Peano requirements are exactly the same as the normal (XRT)

> MLIR-AIE flow — if XRT examples build/run on this box, HRX has what it needs too,

Reviewer 2: ```suggestion
IRON flow — if XRT examples build/run on this box, HRX has what it needs too,
```

`lib/libhrx.so` + `lib/cmake/hrx`); the helper synthesizes an `env.sh` that
exports `HRX_DIR` / `LD_LIBRARY_PATH` / `CMAKE_PREFIX_PATH`, which §4's
auto-detection consumes:

```bash
source "$(hrx-integration/fetch-hrx-release.sh)"
# or, in one step:  eval "$(hrx-integration/fetch-hrx-release.sh --print-env)"
```

This is exactly how the pure-HRX CI job provisions `libhrx`. Once sourced, skip
to §3 — you're done with HRX provisioning. Build from source (§2b) only if you
need a custom `libhrx`.

### 2b. Advanced: build `libhrx` from source

Clone HRX, check out a branch that ships `hrx_amdxdna_executable_create`
(`amdxdna-hal-native-rel` or newer), build it NPU-only, and install the public
distribution.

> **Placement (do this for zero-config detection):** clone HRX as a **sibling of
> `mlir-aie`** into `<parent-of-mlir-aie>/hrx-system` and install to
> `hrx-system/build/hrx-install`. Auto-detection then finds it with no env vars.
> (`~/hrx`, `/opt/hrx`, `/usr/local/hrx` install prefixes also work; anything
> else needs the env hints in §4.)

```bash
# from the directory that contains your mlir-aie checkout:
git clone https://github.com/ROCm/hrx-system.git hrx-system
cd hrx-system

> git remote add jorn https://github.com/jtuyls/hrx.git

Reviewer 2: Should this be `ROCm/hrx` or this fork?

Reviewer 3: Do we need to include the instructions to build from source or can we drop that for now? If not a hard requirement, I would leave it out for now.

Author: Done - dropped the source build instructions for now. The README points at the pinned HRX release the fetch script consumes; we can add source-build docs later if there's demand.

git fetch jorn amdxdna-hal-native-rel
git checkout -B amdxdna-hal-native-rel jorn/amdxdna-hal-native-rel

python dev.py cmake setup
python dev.py cmake configure \
  -DIREE_ROCM_PATH=/opt/rocm -DIREE_ROCM_DEPENDENCY_MODE=package \
  -DCMAKE_INSTALL_LIBDIR=lib \
  -DCMAKE_C_COMPILER=/opt/rocm/llvm/bin/clang \
  -DCMAKE_CXX_COMPILER=/opt/rocm/llvm/bin/clang++ \
  -DCMAKE_ASM_COMPILER=/opt/rocm/llvm/bin/clang \
  -DCMAKE_AR=/opt/rocm/llvm/bin/llvm-ar -DCMAKE_RANLIB=/opt/rocm/llvm/bin/llvm-ranlib \
  -DCMAKE_EXE_LINKER_FLAGS=-fuse-ld=lld -DCMAKE_SHARED_LINKER_FLAGS=-fuse-ld=lld \
  -DCMAKE_MODULE_LINKER_FLAGS=-fuse-ld=lld -DCMAKE_BUILD_TYPE=RelWithDebInfo \
  -DIREE_HAL_DRIVER_AMDGPU=OFF -DIREE_HAL_DRIVER_AMDXDNA=ON
python dev.py cmake build
cmake --install build/cmake --prefix build/hrx-install --component HrxPublicDist
```

> **libstdc++ note:** if the ROCm `clang++` can't find `libstdc++` (e.g. it
> defaults to a gcc install dir that only ships the runtime `.so.6`), point it at
> a full gcc toolchain by adding
> `-DCMAKE_C_FLAGS=--gcc-install-dir=<dir> -DCMAKE_CXX_FLAGS=--gcc-install-dir=<dir>`
> (e.g. `/usr/lib/gcc/x86_64-linux-gnu/13`) to the configure line.

> **Verify** the install produced what the MLIR-AIE HRX backend consumes — the

Reviewer 2: ```suggestion
**Verify** the install produced what the IRON HRX backend consumes — the
```
